### PR TITLE
API: map() on Index returns an Index, not array

### LINKED ISSUE #12766: API: Index.map should return Index rather than array
``` python
df2=pd.DataFrame([[0,1],[2,3]],columns=['c1','c2'],index=['i1','i2'])
df2.index.name='index'
df2.to_html()
```

``` html
<table border="1" class="dataframe">\n  <thead>\n    <tr style="text-align: right;">\n      <th></th>\n      <th>c1</th>\n      <th>c2</th>\n    </tr>\n    <tr>\n      <th>index</th>\n      <th></th>\n      <th></th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>i1</th>\n      <td>0</td>\n      <td>1</td>\n    </tr>\n    <tr>\n      <th>i2</th>\n      <td>2</td>\n      <td>3</td>\n    </tr>\n  </tbody>\n</table>
```

``` python
df2.index = df2.index.map(lambda x: x.upper())
df2.to_html()
```

``` html
<table border="1" class="dataframe">\n  <thead>\n    <tr style="text-align: right;">\n      <th></th>\n      <th>c1</th>\n      <th>c2</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>I1</th>\n      <td>0</td>\n      <td>1</td>\n    </tr>\n    <tr>\n      <th>I2</th>\n      <td>2</td>\n      <td>3</td>\n    </tr>\n  </tbody>\n</table>
```

I think `map` should accept `inplace`

```
INSTALLED VERSIONS
------------------
commit: None
python: 2.7.11.final.0
python-bits: 64
OS: Windows
OS-release: 7
machine: AMD64
processor: Intel64 Family 6 Model 60 Stepping 3, GenuineIntel
byteorder: little
LC_ALL: None
LANG: None

pandas: 0.18.0
nose: 1.3.7
pip: 8.1.1
setuptools: 18.2
Cython: None
numpy: 1.10.4
scipy: 0.17.0
statsmodels: None
xarray: None
IPython: 4.1.2
sphinx: None
patsy: None
dateutil: 2.5.1
pytz: 2016.2
blosc: None
bottleneck: 1.0.0
tables: None
numexpr: 2.5
matplotlib: 1.5.1
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: None
httplib2: None
apiclient: None
sqlalchemy: 1.0.12
pymysql: None
psycopg2: None
jinja2: 2.8
boto: None
```

### PR DESCRIPTION
**Pull request by jrings**
Continued in #14506

- [x] closes #12766 
- [x] tests added / passed
- [x] passes `git diff upstream/master | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

**Comment by jrings** [comment]
Not sure about where the whatsnew entry will go?

File: pandas/indexes/base.py [hunk 1]
```diff
@@ -2198,7 +2198,7 @@ def groupby(self, to_groupby):
         return self._groupby(self.values, _values_from_object(to_groupby))
 
     def map(self, mapper):
-        return self._arrmap(self.values, mapper)
+        return Index(self._arrmap(self.values, mapper))
```
Comment (sinhrks) [inline pandas/indexes/base.py]: This impl losts metadata. Pls use  `_shallow_copy_with_infer`.

File: pandas/tests/indexes/test_base.py [hunk 2]
```diff
@@ -1557,6 +1557,11 @@ def test_string_index_repr(self):
 
                 self.assertEqual(coerce(idx), expected)
 
+    def test_map(self):
+        df = pd.DataFrame([[0, 1], [2, 3]], columns=['c1', 'c2'], index=['i1', 'i2'])
+        df.index = df.index.map(lambda x: x.upper())
+        self.assertTrue(df.index.equals(Index(['I1', 'I2'])))
```
Comment (sinhrks) [inline pandas/tests/indexes/test_base.py]: Pls add following cases:

- dtype changes
- has ``name`` (and ``map`` preserves the ``name``)

**Comment by sinhrks** [comment]
Thanks! whatsnew locates on ``doc/source/whatsnew``.

**Comment by sinhrks** [comment]
Also, I think it is an API change, rather than a bug (as original issue is not categorized).

**Comment by jrings** [comment]
Thanks @sinhrks for the friendly review! This is my first PR, so it's a good opportunity to get a start!
I made those changes, let me know if this is what you had in mind!

Also, the Travis failures seem to be unrelated?

**Comment by jreback** [comment]
@jrings you need to rebase on master. Things changed a bit today.

**Comment by jrings** [comment]
Ugh rebase doesn't like me. Never has never will. Have rebased, will try to squash if changes are ok? (or cry for help)

**Comment by jreback** [comment]
http://pandas.pydata.org/pandas-docs/stable/contributing.html#contributing-your-changes-to-pandas

```
git checkout yourbranch
git rebase -i origin/master
```

**Comment by jrings** [comment]
Yeah I merged on the master to my master, then to my branch. Will work on fixing it, I finally need to figure this out :D

File: ci/requirements-3.5_OSX.run [hunk 3]
```diff
@@ -1,5 +1,5 @@
 pytz
-numpy
+numpy=1.10.4
```
Comment (sinhrks) [inline ci/requirements-3.5_OSX.run]: Can you revert unnecessary changes?

File: doc/source/whatsnew/v0.18.1.txt [hunk 4]
```diff
@@ -103,6 +103,11 @@ API changes
 - ``CParserError`` is now a ``ValueError`` instead of just an ``Exception`` (:issue:`12551`)
 
 - ``pd.show_versions()`` now includes ``pandas_datareader`` version (:issue:`12740`)
+- Provide a proper ``__name__`` and ``__qualname__`` attributes for generic functions (:issue:`12021`)
+
+- ``map`` on an ``Index`` now returns an ``Index``, not an array (:issue:`12766`)
+
+- ``map`` on an ``Index`` now returns an ``Index``, not an array (:issue:`12766`)
```
Comment (sinhrks) [inline doc/source/whatsnew/v0.18.1.txt]: Also here.

**Comment by sinhrks** [comment]
After properly rebased, pls fix datetimelike `map` consistently. 

- https://github.com/pydata/pandas/blob/master/pandas/tseries/base.py#L226

Also, adding more dtype tests (you can find some examples here).

- https://github.com/pydata/pandas/blob/master/pandas/tests/indexes/test_base.py#L37

File: ci/requirements-3.5_OSX.run (hunk 3 above)
Comment (jrings) [inline ci/requirements-3.5_OSX.run]: git blame tells me this comes from @jreback https://github.com/jrings/pandas/commit/fded94274effa2d592e48ec2079ba3b78e5bb232

Comment (sinhrks) [inline ci/requirements-3.5_OSX.run]: Pls only include changes you've made(see other PRs). You need rebase:

git checkout master
git pull upstream
git checkout <your branch>
git rebase -i master

**Comment by jrings** [comment]
Ok, cleaned up the rebase mess, will react to the rest later or tomorrow. Thanks!

**Comment by jrings** [comment]
@sinhrks Oh I see, by different dtypes you meant testing map on the different index flavors, not a type conversion on the Index?

**Comment by sinhrks** [comment]
@jrings thx for the fix. yes i meant different classes.

**Comment by jrings** [comment]
Ok, need some advice. Added almost all the index types successfully, but failed on the categorical:

```
======================================================================
ERROR: test_map (pandas.tests.indexes.test_base.TestIndex)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/joerg/src/pandas/pandas/tests/indexes/test_base.py", line 1593, in test_map
    testIdx = self.catIndex.map(lambda x: int(x))
  File "/home/joerg/src/pandas/pandas/indexes/base.py", line 2201, in map
    return self._shallow_copy_with_infer(self._arrmap(self.values, mapper))
TypeError: Argument 'index' has incorrect type (expected numpy.ndarray, got Categorical)
```

Can you give me a pointer to how to fix this?

File: doc/source/whatsnew/v0.18.1.txt [hunk 5]
```diff
@@ -105,6 +105,10 @@ API changes
 - ``pd.show_versions()`` now includes ``pandas_datareader`` version (:issue:`12740`)
 - Provide a proper ``__name__`` and ``__qualname__`` attributes for generic functions (:issue:`12021`)
 
+- ``map`` on an ``Index`` now returns an ``Index``, not an array (:issue:`12766`)
+
+- ``map`` on an ``Index`` now returns an ``Index``, not an array (:issue:`12766`)
```
Comment (sinhrks) [inline doc/source/whatsnew/v0.18.1.txt]: dupe.

File: pandas/tests/indexes/test_base.py [hunk 6]
```diff
@@ -1557,6 +1557,46 @@ def test_string_index_repr(self):
...
+        testIdx = self.floatIndex.map(lambda x: -1 if x < 0 else 1)
+        self.assertEqual(len(testIdx), 100)
+        self.assertTrue(isinstance(testIdx, Int64Index))
+        self.assertTrue(set(testIdx == {-1, 1}))
+
+        testIdx = self.boolIndex.map(lambda x: not x)
+        self.assertTrue(testIdx.equals(Index([False, True])))
+
+        testIdx = self.catIndex.map(lambda x: int(x))
+        print(testIdx)
```
Comment (sinhrks) [inline pandas/tests/indexes/test_base.py]: Please remove. Also, use ``tm.assert_index_equal`` for ``Index`` comparison.

**Comment by sinhrks** [comment]
The error should be handled in #12532. Let me update it.

**Comment by jrings** [comment]
Ok, yeah sorry that `print` was left over from figuring out the catIndex, and I'll dedupe the whatsnew entry as well. Should I wait for #12532 to be merged then?

**Comment by jrings** [comment]
Also, should I do anything for empty and tuples Index?

**Comment by sinhrks** [comment]
#12532 has been merged, can you update?

File: pandas/indexes/base.py [hunk 7]
```diff
@@ -2198,7 +2198,20 @@ def groupby(self, to_groupby):
         return self._groupby(self.values, _values_from_object(to_groupby))
 
     def map(self, mapper):
-        return self._arrmap(self.values, mapper)
+        """
+        Apply mapper function to an index
+
+         Parameters
```
Comment (jorisvandenbossche) [inline pandas/indexes/base.py]: The indentation is one space off here (as well as all lines below)

Comment (jrings) [inline pandas/indexes/base.py]: Corrected!

**Comment by jrings** [comment]
Ok categorical works now! Rebased on master and squashed commits.
@sinhrks

File: pandas/tests/indexes/test_base.py [hunk 8]
```diff
@@ -1557,6 +1557,41 @@ def test_string_index_repr(self):
 
                 self.assertEqual(coerce(idx), expected)
 
+    def test_map(self):
```
Comment (jreback) [inline pandas/tests/indexes/test_base.py]: put all of this in ``pandas/tests/index/common.py/Base`` it will then get tested on each index type. see the structure of other methods. If you need to override it then put it specifically in a ``pandas/test/index/test_*`` file.

**Comment by jrings** [comment]
Will work on it this weekend!

**Comment by jrings** [comment]
Sorry for being slow, am traveling a lot at the moment. But one question @jreback, would a very simple function for `map` be enough, just to test that a map still produces an `Index`, because then putting it all in the loop of the various indexes would be fine, but if I would try to transfer my more varied tests that would lead to a lot of `isinstance` for each type that would sort of defeat the purpose - so then I'd distribute them over all the different test classes for each type of index?

File: doc/source/whatsnew/v0.18.1.txt [hunk 9]
```diff
@@ -149,6 +149,8 @@ API changes
 - Provide a proper ``__name__`` and ``__qualname__`` attributes for generic functions (:issue:`12021`)
 - ``pd.concat(ignore_index=True)`` now uses ``RangeIndex`` as default (:issue:`12695`)
 
+- ``map`` on an ``Index`` now returns an ``Index``, not an array (:issue:`12766`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.18.1.txt]: move to 0.18.2

File: pandas/tests/indexes/test_base.py [hunk 10]
```diff
@@ -1557,6 +1557,41 @@ def test_string_index_repr(self):
...
+
+        # Applying a function to the Index
+        df = pd.DataFrame([[0, 1], [2, 3]], columns=['c1', 'c2'], index=['i1', 'i2'])
+        df.index.name = "Numbering"
+        df.index = df.index.map(lambda x: x.upper())
```
Comment (jreback) [inline pandas/tests/indexes/test_base.py]: you need to move the tests to ``pandas/tests/index/common.py/Base`` as I indicated above. Then you can follow the other test examples; it will be then run for each index type.

**Comment by jreback** [comment]
can you rebase/update

File: pandas/tests/indexes/test_base.py (hunk 10 above)
Comment (jrings) [inline pandas/tests/indexes/test_base.py]: Ok finally able to get back to this. Will a very simple test be enough than (like -> map with identity function returns an Index)? Otherwise it'll need a lot of `if isinstance`...

Comment (jreback) [inline pandas/tests/indexes/test_base.py]: see my comments above these need to go in common.py

**Comment by jreback** [comment]
can you rebase/update

**Comment by jreback** [comment]
can you rebase / update?

**Comment by jreback** [comment]
pls rebase / update

**Comment by nateyoder** [comment]
If this has stalled I'd be more than happy to take it over as I'd really like to see and #12756 make it into a release soon.  Is the easiest way to do this by creating a new PR or can I use this one somehow?

**Comment by jreback** [comment]
create a new PR. you can incorporate this commit (if you'd like). you can add a closes (this PR number) as well as the issue number.

**Comment by jorisvandenbossche** [comment]
Continued in #14506
